SignalServiceKit/src/Util/DataSource.m

```diff
@@ -0,0 +1,280 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import "DataSource.h"
+#import "MIMETypeUtil.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface DataSourceValue ()
+
+@property (nonatomic) NSData *dataValue;
+
+@property (nonatomic) NSString *fileExtension;
+
+// This property is lazy-populated.
+@property (nonatomic) NSString *cachedFilePath;
+
+@property (nonatomic, nullable) NSString *sourceFilename;
+
+@end
+
+#pragma mark -
+
+@implementation DataSourceValue
+
++ (nullable id<DataSource>)dataSourceWithData:(NSData *)data fileExtension:(NSString *)fileExtension
+{
+    OWSAssert(data);
+
+    if (!data) {
+        return nil;
+    }
+
+    DataSourceValue *instance = [DataSourceValue new];
+    instance.dataValue = data;
+    instance.fileExtension = fileExtension;
+    return instance;
+}
+
++ (nullable id<DataSource>)dataSourceWithData:(NSData *)data utiType:(NSString *)utiType
+{
+    NSString *fileExtension = [MIMETypeUtil fileExtensionForUTIType:utiType];
+    return [self dataSourceWithData:data fileExtension:fileExtension];
+}
+
++ (nullable id<DataSource>)dataSourceWithOversizeText:(NSString *_Nullable)text
+{
+    OWSAssert(text);
```

Comment:
nit: This assert seems to contradict the nullability annotation of the `text` param. In other words, should this declaration be: 

```
+ (nullable id<DataSource>)dataSourceWithOversizeText:(NSString *)text
```

So that the compiler can catch places where we might pass in a nil `text`?

Reply:
I've made the text non-null.